File: lint-arwaky/src-python/infrastructure/memory_registry_adapter.py
"""Infrastructure adapter for in-memory job tracking."""

from __future__ import annotations

import uuid
import asyncio
import logging

from datetime import datetime, timezone

from ..taxonomy import (
    Count,
    JobId,
    ActionName,
    SuccessStatus,
    ErrorMessage,
    MetadataVO,
    Identity,
    Duration,
    ResponseData,
)
from ..contract import IJobRegistryPort


class MemoryJobRegistryAdapter(IJobRegistryPort):
    """Implementation of job tracking and lifecycle management."""

    def __init__(self) -> None:
        super().__init__()
        self._jobs: dict = {}
        self._lock: asyncio.Lock = asyncio.Lock()
        self.logger = logging.getLogger("infra.jobs")

    def _get_lock(self) -> asyncio.Lock:
        return self._lock

    async def create_job(self, action: ActionName | Identity) -> JobId:
        """Register a new job and return its ID."""
        action_str = str(action.value) if hasattr(action, "value") else str(action)
        job_id_str = str(uuid.uuid4())[:8]
        job_id = JobId(value=job_id_str)
        async with self._get_lock():
            self._jobs[job_id_str] = {
                "id": job_id_str,
                "status": "running",
                "action": action_str,
                "started_at": datetime.now(timezone.utc).isoformat(),
            }
        return job_id

    async def complete_job(
        self, job_id: JobId | Identity, result: ResponseData | MetadataVO
    ):
        """Mark job as completed."""
        jid = str(job_id.value) if hasattr(job_id, "value") else str(job_id)
        async with self._get_lock():
            if jid in self._jobs:
                self._jobs[jid]["status"] = "completed"
                self._jobs[jid]["result"] = result
                self._jobs[jid]["completed_at"] = datetime.now(timezone.utc).isoformat()

    async def fail_job(
        self, job_id: JobId | Identity, error: ErrorMessage | Identity
    ) -> None:
        """Mark job as failed."""
        jid = str(job_id.value) if hasattr(job_id, "value") else str(job_id)
        error_str = str(error.value) if hasattr(error, "value") else str(error)
        async with self._get_lock():
            if jid in self._jobs:
                self._jobs[jid]["status"] = "failed"
                self._jobs[jid]["error"] = error_str
                self._jobs[jid]["failed_at"] = datetime.now(timezone.utc).isoformat()

    async def list_jobs(self) -> MetadataVO:
        """Return all jobs."""
        async with self._get_lock():
            return MetadataVO(data={"jobs": list(self._jobs.values())})

    async def get_job(self, job_id: JobId | Identity) -> MetadataVO | None:
        """Return a single job or None."""
        jid = str(job_id.value) if hasattr(job_id, "value") else str(job_id)
        async with self._get_lock():
            job = self._jobs.get(jid)
            if not job:
                return None
            return MetadataVO(data=job)

    async def cancel_job(self, job_id: JobId | Identity) -> SuccessStatus:
        """Cancel a running job. Returns SuccessStatus(True) if cancelled."""
        jid = str(job_id.value) if hasattr(job_id, "value") else str(job_id)
        async with self._get_lock():
            if jid in self._jobs:
                self._jobs[jid]["status"] = "cancelled"
                self._jobs[jid]["completed_at"] = datetime.now(timezone.utc).isoformat()
                return SuccessStatus(status=True)
            return SuccessStatus(status=False)

    async def run_with_retry(
        self,
        operation: Identity,
        max_retries: Count = Count(value=5),
        base_delay: Duration = Duration(value=0.5),
    ) -> ResponseData:
        """Execute an operation with exponential backoff retry logic.

        Note: This is a mock implementation.
        """
        return ResponseData(content=f"Executed {operation}")

File: lint-arwaky/src-rust/infrastructure/memory_registry_adapter.rs
/// memory_registry_adapter — In-memory job tracking implementation.
use crate::contract::IJobRegistryPort;
use crate::taxonomy::{Count, Duration, ErrorMessage, Identity, JobId, MetadataVO, ResponseData, SuccessStatus, BooleanVO, JobError, JobStatus};
use std::collections::HashMap;
use tokio::sync::Mutex;

#[derive(Debug, Clone)]
struct JobRecord {
    id: String,
    status: String,
    action: String,
    started_at: String,
    result: Option<String>,
    error: Option<String>,
    completed_at: Option<String>,
}

pub struct MemoryJobRegistryAdapter {
    jobs: Mutex<HashMap<String, JobRecord>>,
}

impl MemoryJobRegistryAdapter {
    pub fn new() -> Self {
        Self { jobs: Mutex::new(HashMap::new()) }
    }

    pub async fn create_job(&self, action: &str) -> JobId {
        let job_id = format!("{:x}", rand::random::<u32>());
        let record = JobRecord {
            id: job_id.clone(),
            status: "running".to_string(),
            action: action.to_string(),
            started_at: chrono::Utc::now().to_rfc3339(),
            result: None,
            error: None,
            completed_at: None,
        };
        self.jobs.lock().await.insert(job_id.clone(), record);
        JobId::new(job_id)
    }

    pub async fn complete_job(&self, job_id: &JobId, result: &ResponseData) {
        let mut jobs = self.jobs.lock().await;
        if let Some(record) = jobs.get_mut(&job_id.value) {
            record.status = "completed".to_string();
            record.result = Some(result.value.to_string());
            record.completed_at = Some(chrono::Utc::now().to_rfc3339());
        }
    }

    pub async fn fail_job(&self, job_id: &JobId, error: &ErrorMessage) {
        let mut jobs = self.jobs.lock().await;
        if let Some(record) = jobs.get_mut(&job_id.value) {
            record.status = "failed".to_string();
            record.error = Some(error.value.clone());
            record.completed_at = Some(chrono::Utc::now().to_rfc3339());
        }
    }

    pub async fn list_jobs(&self) -> Vec<serde_json::Value> {
        let jobs = self.jobs.lock().await;
        jobs.values().map(|j| serde_json::json!({
            "id": j.id,
            "status": j.status,
            "action": j.action,
            "started_at": j.started_at,
            "completed_at": j.completed_at,
        })).collect()
    }

    pub async fn cancel_job(&self, job_id: &JobId) -> SuccessStatus {
        let mut jobs = self.jobs.lock().await;
        if let Some(record) = jobs.get_mut(&job_id.value) {
            record.status = "cancelled".to_string();
            record.completed_at = Some(chrono::Utc::now().to_rfc3339());
            SuccessStatus::new(BooleanVO::new(true))
        } else {
            SuccessStatus::new(BooleanVO::new(false))
        }
    }
}

#[async_trait::async_trait]
impl IJobRegistryPort for MemoryJobRegistryAdapter {
    async fn create_job(&self, action: &str) -> Result<JobId, JobError> {
        unimplemented!()
    }
    async fn complete_job(&self, job_id: &JobId, result: &ResponseData) {
        unimplemented!()
    }
    async fn fail_job(&self, job_id: &JobId, error: &ErrorMessage) {
        unimplemented!()
    }
    async fn list_jobs(&self) -> Vec<serde_json::Value> {
        unimplemented!()
    }
    async fn get_job(&self, job_id: &JobId) -> Option<JobId> {
        unimplemented!()
    }
    async fn cancel_job(&self, job_id: &JobId) -> SuccessStatus {
        unimplemented!()
    }
    async fn run_with_retry(
        &self,
        operation: &str,
        max_retries: u32,
        base_delay: Duration,
    ) -> ResponseData {
        unimplemented!()
    }
}

